src/commands/branch.rs

```diff
@@ -61,7 +61,14 @@ pub struct BranchDeleteArgs {
 /// preceded by a "+". For information about branches, see
 /// https://github.com/martinvonz/jj/blob/main/docs/branches.md.
 #[derive(clap::Args, Clone, Debug)]
-pub struct BranchListArgs;
+pub struct BranchListArgs {
+    /// Show branches which local/remote targets are in the given revisions.
```

Comment:
Perhaps this should, by default, match what `jj git push -r` pushes? Then the comment could mention that.

That would mean not including remote-tracking branches by default, though we'd probably want to add an option to include them.

Reply:
Interesting. I don't consider `jj branch list` in push/fetch context. In your workflow, is `jj branch list` a preflight command for push?

> That would mean not including remote-tracking branches by default, though we'd probably want to add an option to include them.

Yeah, we'll need separate option if `-r` is restricted to local targets. It might have to be per-remote. For queries like `-r 'branches()'`, I don't think we'll need such strictness. In order to preview outgoing branches, `jj branch list -r` must be strict.

Reply:
I think it could be a nice program command, but I haven't tried out `push -r` much. Lately, I've been doing `push --all` mostly.

I might just need to look at the tests again, but what's the use case for using -r on a revision that a remote branch points to?

I'll probably won't look at this again until tomorrow.

Reply:
> use case for using -r on a revision that a remote branch points to?

`-r 'remote_branch(remote=foo)'`

Well, I personally wouldn't want to query such things. Perhaps, more important property is that `-r 'all()'` includes deleted/forgotten branches. If we had `revsets.branch_list` config knob, we would need something to cancel out. It could be OR-ed `--deleted --forgotten` flags, but I don't know.

Reply:
> If we had `revsets.branch_list` config knob, we would need something to cancel out. It could be OR-ed `--deleted --forgotten` flags,

OR-ed `--deleted` seems bad idea since it's unclear whether each option is AND-ed or OR-ed. For `--glob NAME --deleted`, I would expect `--glob NAME` AND `--deleted`.

Perhaps, exclusive `--all` flag is better to disable default filtering.

Reply:
> Perhaps, more important property is that `-r 'all()'` includes deleted/forgotten branches.

I'm worried that might not work anyway.

Let's say you fetch a branch, change it's description, and then delete it. Now, it's only trace is a remote-tracking branch on an abandoned commit. Will `-r 'all()'` pick it up?

I keep hoping to rename all() into alive() or something, but I don't like the name visible() am not sure alive/dead is good. active/obsolete seems nice to me, but those words are long.

Reply:
Sorry for a slow reply.

> I lean towards restricting the query to local targets (consistency is good), and the only pain point is that there's no way to populate deleted/forgotten branches if we had a default revset. I'm not gonna add --all and default revset in this PR, though.

This sounds best to me. I think supporting globs would be pretty good for showing deleted/forgotten branches. As I elaborate on below, I think the way jj's revsets presently interact with the notion of abandoned commits is quite confusing, so I wouldn't be very comfortable using `-r` when I want to refer to abandoned commits anyway.

> It does in this context. 

I am not sure how relevant this point is anymore, but even though our test proves that it does, it's quite surprising to me. If `jj log -r all()` skips un-abandoned commits, so should this. I clearly don't understand something about how revsets work.

> BTW, do you like the idea of adding --all flag to disable default filtering (if any)?

(This is also no longer very relevant for this PR, I think.)

I didn't really understand this idea. I think that's connected to my confusion about the previous point. In my mental model, if we're doing revsets, they should specify an exact set of commits, and not be subject to any further filtering (even if jj doesn't seem to follow this perfectly right now). There might be value in changing this model and have some notion of different filtering for the same revset, but it's not the easiest task to create a model that does that and isn't overly confusing. I can't think of a way of understanding the `-r all()` issue from above that's not confusing.

Reply:
Ok, let's restrict `-r` query to local_targets.

> If `jj log -r all()` skips un-abandoned commits, so should this.

Off-topic: The problem in `log` is that user can populate hidden revisions by explicitly specifying them, and the meaning of `all()` would get fuzzier.

> In my mental model, if we're doing revsets, they should specify an exact set of commits,

No, they aren't for my use case. My short-term goal is "hiding remote-only branches by default" #1136. It seems also useful to filter branches by `-r'author(me)'`.

Suppose user added `revsets.branch_list = "author(me)"` to config file, `jj branch list` won't show any other branches by default. We'll need some way to override the default, but `-r 'all()'` wouldn't work because it can't select deleted/forgotten branches. So, `--all` flag will disable the default revset. `--all -r` or `--all --glob` should be banned.

Reply:
> My short-term goal is "hiding remote-only branches by default" https://github.com/martinvonz/jj/issues/1136. It seems also useful to filter branches by -r'author(me)'.

Ah, that's an interesting answer to my question about "what's the use case".

I would really like there to be some way to distinguish between the revset of "unabandoned commits authored by me" and "all commits authored by me" (perhaps with even more possibilities, e.g. "commits from my default revsets authored by me). Ideally, we'd solve that, and the solution would apply here as well.

Until then, I think something like `--all` would be a decent workaround. Perhaps we could call it `--include-deleted`, then I think it's pretty clear that `--include-deleted --glob` shows more branches than just `--glob`. I think it's OK for `--include-deleted` to actually mean `--include-deleted-or-forgotten`.

Reply:
> the revset of "unabandoned commits authored by me" and "all commits authored by me" (perhaps with even more possibilities, e.g. "commits from my default revsets authored by me). Ideally, we'd solve that, and the solution would apply here as well.

OT, but to my knowledge:
* unabandoned commits authored by me: `:visible_heads() & author(me)`
* all commits authored by me: `author(me)` (but "all" here is unclear. it's unlikely I would want to see tons of hidden commits)
* commits from my default revsets authored by me: we'll somehow need to unify `[revsets]` and `[revset-aliases]`)

Anyway, none of them can't populate deleted/forgotten branches since their `local_target` is `None`.

Reply:
How do we reconcile this with the fact that `jj log -r 'author(me)'` doesn't show abandoned commits?

If it's temporary, I think it's not too bad for `jj branch list` to silently have a different behavior and apply revsets to abandoned commits as well. For the purposes of this PR, it's probably the best option.

But I think we should have a better resolution of this discrepancy long-term, if at all possible.

Reply:
> How do we reconcile this with the fact that `jj log -r 'author(me)'` doesn't show abandoned commits?

I don't see why that can be a problem. The initial set to be filtered by revset is different. `jj branch list` wouldn't show non-branch-head commits.

Anyway, it's unlikely that a local branch ref points to abandoned commit. I showed it's doable in tests, but that might be actually a bug that `jj branch set` doesn't re-add the target commit to heads.

Reply:
> Anyway, it's unlikely that a local branch ref points to abandoned commit. I showed it's doable in tests, but that might be actually a bug that jj branch set doesn't re-add the target commit to heads.

You're right, and I think `jj branch set` should make the branch visible.

> I don't see why that can be a problem. The initial set to be filtered by revset is different.

As you said above, it doesn't matter very much for this PR.

But, before reviewing this PR, I was sure that `all()` is defined as the set of all visible commits. Partially, it is because the word `revset` includes the word `set`. I suppose I can think of them as `revfilter`s. Even then, I think this notion of "the initial set" should be more explicit.

Reply:
If you're interested in weird part of `all()`, see
https://github.com/martinvonz/jj/blob/335e86d8e756188cb5eb354f18af816b51d28d5e/lib/src/revset.rs#L2047-L2059

Reply:
🥲